multisig2/src/lib.rs

```diff
@@ -0,0 +1,878 @@
+}
+
+#[inline]
+fn assert(condition: bool, error: &str) {
+    if !condition {
+        env::panic_str(error);
+    }
+}
+
+#[near_bindgen]
+impl MultiSigContract {
+    /// Initialize multisig contract.
+    /// @params members: list of {"account_id": "name"} or {"public_key": "key"} members.
+    /// @params num_confirmations: k of n signatures required to perform operations.
+    #[init]
+    pub fn new(members: Vec<MultisigMember>, num_confirmations: u32) -> Self {
+        assert(
+            members.len() >= num_confirmations as usize,
+            "Members list must be equal or larger than number of confirmations",
+        );
+        let mut multisig = Self {
+            members: UnorderedSet::new(StorageKeys::Members),
+            num_confirmations,
+            request_nonce: 0,
+            requests: UnorderedMap::new(StorageKeys::Requests),
+            confirmations: LookupMap::new(StorageKeys::Confirmations),
+            num_requests_pk: LookupMap::new(StorageKeys::NumRequestsPk),
+            active_requests_limit: ACTIVE_REQUESTS_LIMIT,
+        };
+        let mut promise = Promise::new(env::current_account_id());
+        for member in members {
+            promise = multisig.add_member(promise, member);
+        }
+        multisig
+    }
+
+    /// Add request for multisig.
+    pub fn add_request(&mut self, request: MultiSigRequest) -> RequestId {
+        let current_member = self.current_member().unwrap_or_else(|| {
+            env::panic_str(
+                "Predecessor must be a member or transaction signed with key of given account",
+            )
+        });
+        // track how many requests this key has made
+        let num_requests = self
+            .num_requests_pk
+            .get(&current_member.to_string())
+            .unwrap_or(0)
+            + 1;
+        assert(
+            num_requests <= self.active_requests_limit,
+            "Account has too many active requests. Confirm or delete some.",
+        );
+        self.num_requests_pk
+            .insert(&current_member.to_string(), &num_requests);
+        // add the request
+        let request_added = MultiSigRequestWithSigner {
+            member: current_member,
+            added_timestamp: env::block_timestamp(),
+            request,
+        };
+        self.requests.insert(&self.request_nonce, &request_added);
+        let confirmations = HashSet::new();
+        self.confirmations
+            .insert(&self.request_nonce, &confirmations);
+        self.request_nonce += 1;
+        self.request_nonce - 1
+    }
+
+    /// Add request for multisig and confirm with the pk that added.
+    pub fn add_request_and_confirm(&mut self, request: MultiSigRequest) -> RequestId {
+        let request_id = self.add_request(request);
+        self.confirm(request_id);
+        request_id
+    }
+
+    /// Remove given request and associated confirmations.
+    pub fn delete_request(&mut self, request_id: RequestId) {
```

Comment:
Might not be a bad idea to return `MultiSigRequest` since `remove_request` will give us that anyway.
Might be useful for a frontend to say, "Deleted <some info>…"
Not strongly held, just a thought.